HEN_HOUSE/egs++/ausgab_objects/egs_phsp_scoring/egs_phsp_scoring.cpp

```diff
@@ -0,0 +1,718 @@
+/*
+###############################################################################
+#
+#  EGSnrc egs++ phsp scoring object
+#  Copyright (C) 2015 National Research Council Canada
+#
+#  This file is part of EGSnrc.
+#
+#  EGSnrc is free software: you can redistribute it and/or modify it under
+#  the terms of the GNU Affero General Public License as published by the
+#  Free Software Foundation, either version 3 of the License, or (at your
+#  option) any later version.
+#
+#  EGSnrc is distributed in the hope that it will be useful, but WITHOUT ANY
+#  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+#  FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for
+#  more details.
+#
+#  You should have received a copy of the GNU Affero General Public License
+#  along with EGSnrc. If not, see <http://www.gnu.org/licenses/>.
+#
+###############################################################################
+#
+#  Author:          Blake Walters, 2018
+#
+###############################################################################
+#
+#  A phase space scoring class.
+#
```

Comment:
Hi @rtownson:  One of my original ideas for this ausgab object was to use the setLabels function to determine whether particles are "inside" or "outside" of the scoring geometry, but I couldn't figure out a way to retroactively go back and set labels in the transport geometry once it had already been created.  Can you see a way to do this?  Even if region labeling can't be used, though, it should be no problem to score based on entering/exiting certain region numbers provided that the user knows the region numbers in the original geometry.  As for multiple passers, yes, in the current implementation, there's no way to tell if a particle has crossed a scoring plane more than once.  I can add an an option to: 1) eliminate particles once they've been scored, 2) flag multiple passers using LATCH in EGSnrc format phase space files or 3) simply not score > 1 pass of a given particle.  Finally, before merging this feature, yes, I'll transfer the description over to the .h file and format it for doxygen.

Reply:
I'm not sure I understand your `setLabels` idea, but could you use `isInside` on the scoring geometry for this? That could work if you want to score upon entering or exiting a given geometry I think. For the idea where the user provides a list of regions (or labels), then you can use labels the same as in egs_dose_scoring.

For multiple passers, it's always good to have options I guess I just don't understand the application of (3). I get confused by the creation of secondaries after the primary is scored - the primary may pass a second time and not get scored but create more secondaries that are scored. Then when you use the phase-space you would have the primary where it was first scored but also secondaries from transport that "hasn't occurred yet". So if you re-ran the simulation from the phase-space you would get a higher dose in the scoring region that in the original simulation.

Maybe there should be an option to not multiple-score the primary *and* its secondaries - this would allow you to continue transport after phase-space scoring (to score multiple phase-spaces for example) but not have signal duplication.

Reply:
@rtownson, see the latest commit.